internal/grpc/services/gateway/storageprovider.go

```diff
@@ -141,16 +141,93 @@ func (s *svc) InitiateFileDownload(ctx context.Context, req *provider.InitiateFi
 		return s.initiateFileDownload(ctx, req)
 	}
 
-	if s.isSharedFolder(ctx, p) || s.isShareName(ctx, p) {
-		log.Debug().Msgf("path:%s points to shared folder or share name", p)
-		err := errtypes.PermissionDenied("gateway: cannot download share folder or share name: path=" + p)
+	if s.isSharedFolder(ctx, p) {
+		log.Debug().Msgf("path:%s points to shared folder", p)
+		err := errtypes.PermissionDenied("gateway: cannot download share folder: path=" + p)
 		log.Err(err).Msg("gateway: error downloading")
 		return &gateway.InitiateFileDownloadResponse{
-			Status: status.NewInvalidArg(ctx, "path points to share folder or share name"),
+			Status: status.NewInvalidArg(ctx, "path points to share folder"),
 		}, nil
 
 	}
 
+	if s.isShareName(ctx, p) {
+		statReq := &provider.StatRequest{Ref: req.Ref}
+		statRes, err := s.stat(ctx, statReq)
+		if err != nil {
+			return &gateway.InitiateFileDownloadResponse{
+				Status: status.NewInternal(ctx, err, "gateway: error stating ref:"+statReq.Ref.String()),
+			}, nil
+		}
+		if statRes.Status.Code != rpc.Code_CODE_OK {
+			if statRes.Status.Code == rpc.Code_CODE_NOT_FOUND {
+				return &gateway.InitiateFileDownloadResponse{
+					Status: status.NewNotFound(ctx, "gateway: file not found:"+statReq.Ref.String()),
+				}, nil
+			}
+			err := status.NewErrorFromCode(statRes.Status.Code, "gateway")
+			return &gateway.InitiateFileDownloadResponse{
+				Status: status.NewInternal(ctx, err, "gateway: error stating ref:"+statReq.Ref.String()),
+			}, nil
+		}
+
+		if statRes.Info.Type != provider.ResourceType_RESOURCE_TYPE_REFERENCE {
+			err := errors.New(fmt.Sprintf("gateway: expected reference: got:%+v", statRes.Info))
+			log.Err(err).Msg("gateway: error stating share name")
+			return &gateway.InitiateFileDownloadResponse{
+				Status: status.NewInternal(ctx, err, "gateway: error initiating download"),
+			}, nil
+		}
+
+		ri, protocol, err := s.checkRef(ctx, statRes.Info)
+		if err != nil {
+			if _, ok := err.(errtypes.IsNotFound); ok {
+				return &gateway.InitiateFileDownloadResponse{
+					Status: status.NewNotFound(ctx, "gateway: reference not found:"+statRes.Info.Target),
+				}, nil
+			}
+			log.Err(err).Msg("gateway: error resolving reference")
+			return &gateway.InitiateFileDownloadResponse{
+				Status: status.NewInternal(ctx, err, "error initiating download"),
+			}, nil
+		}
+		// if it is a file allow download
+		if ri.Type == provider.ResourceType_RESOURCE_TYPE_FILE {
```

Comment:
optional: for the sake of readability, would be nice to have all `if` statement lead to error case branching, and the non-branching code path being the success path